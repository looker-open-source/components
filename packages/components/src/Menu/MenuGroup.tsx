/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import React, { FC, ReactNode, useContext, useState } from 'react'
import styled from 'styled-components'
import { CompatibleHTMLProps, reset } from '@looker/design-tokens'
import { List } from '../List'
import { MenuItemContext } from './MenuContext'
import { MenuGroupLabel } from './MenuGroupLabel'

export interface MenuGroupProps
  extends Omit<CompatibleHTMLProps<HTMLElement>, 'label'> {
  compact?: boolean
  label?: ReactNode
  className?: string
}

const MenuGroupLayout: FC<MenuGroupProps> = ({
  children,
  className,
  compact,
  label,
}) => {
  const [renderIconPlaceholder, setRenderIconPlaceholder] = useState(false)
  const {
    compact: contextCompact,
    renderIconPlaceholder: contextRenderIconPlaceholder,
    ...restContext
  } = useContext(MenuItemContext)

  const context = {
    ...restContext,
    compact: compact === undefined ? contextCompact : compact,
    renderIconPlaceholder:
      contextRenderIconPlaceholder || renderIconPlaceholder,
    setRenderIconPlaceholder,
  }

  return (
    <li className={className}>
      <MenuItemContext.Provider value={context}>
        {label && <MenuGroupLabel>{label}</MenuGroupLabel>}
        <List>{children}</List>
      </MenuItemContext.Provider>
    </li>
  )
}

export const MenuGroup = styled(MenuGroupLayout)`
  ${reset}
  list-style-type: none;
  padding: ${({ theme: { space } }) => space.xsmall} 0;
`
