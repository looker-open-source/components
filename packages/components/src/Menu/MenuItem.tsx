/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import {
  CompatibleHTMLProps,
  size,
  space,
  SizeProps,
  SpaceProps,
} from '@looker/design-tokens'
import { IconNames } from '@looker/icons'
import styled from 'styled-components'
import React, { FC, ReactNode, useContext, useState, useEffect } from 'react'
import { Placement } from '@popperjs/core'
import { Paragraph } from '../Text'
import { useID } from '../utils/useID'
import { Icon } from '../Icon'
import { Tooltip } from '../Tooltip'
import { MenuContext, MenuItemContext } from './MenuContext'
import { MenuItemLayout } from './MenuItemLayout'

export interface MenuItemProps extends CompatibleHTMLProps<HTMLElement> {
  iconArtwork?: ReactNode
  compact?: boolean
  /**
   * Indicates the MenuItem is checked
   */
  current?: boolean
  /*
   * optional extra description
   */
  description?: ReactNode
  detail?: ReactNode
  icon?: IconNames
  /**
   * Sets the correct accessible role for the MenuItem:
   * - Use **'link'** for items that navigation to another page
   * - Use **'button'** for items that trigger in page interactions, like displaying a dialog
   * @default 'button'
   *
   */
  itemRole?: 'link' | 'button'
  tooltip?: string
  tooltipPlacement?: Placement
}

const MenuItemInternal: FC<MenuItemProps> = (props) => {
  const {
    children,
    className,
    compact: propCompact,
    current,
    description,
    detail,
    disabled,
    href,
    icon,
    iconArtwork,
    itemRole,
    onBlur,
    onClick,
    onKeyUp,
    target,
    tooltip,
    tooltipPlacement,
  } = props

  const [isFocusVisible, setFocusVisible] = useState(false)
  const { compact: contextCompact } = useContext(MenuItemContext)
  const compact = propCompact === undefined ? contextCompact : propCompact

  const handleOnBlur = (event: React.FocusEvent<HTMLLIElement>) => {
    setFocusVisible(false)
    onBlur && onBlur(event)
  }

  const { setOpen } = useContext(MenuContext)
  const {
    renderIconPlaceholder,
    setRenderIconPlaceholder,
    handleArrowDown,
    handleArrowUp,
  } = useContext(MenuItemContext)

  const handleOnClick = (event: React.MouseEvent<HTMLLIElement>) => {
    setFocusVisible(false)
    onClick && onClick(event)
    // Close the Menu (unless event has preventDefault in onClick)
    if (setOpen && !event.defaultPrevented) {
      setOpen(false)
    }
  }

  const handleOnKeyDown = (event: React.KeyboardEvent<HTMLLIElement>) => {
    switch (event.key) {
      case 'ArrowUp':
        handleArrowUp && handleArrowUp(event)
        break
      case 'ArrowDown':
        handleArrowDown && handleArrowDown(event)
        break
    }
  }

  const handleOnKeyUp = (event: React.KeyboardEvent<HTMLLIElement>) => {
    onKeyUp && onKeyUp(event)
    setFocusVisible(true)
  }

  useEffect(() => {
    icon && setRenderIconPlaceholder(true)
  }, [icon, setRenderIconPlaceholder])

  const renderedIconID = useID(props.id)

  const renderedIcon =
    icon || iconArtwork ? (
      <Icon
        artwork={iconArtwork}
        color="text1"
        name={icon}
        size={24 / (compact ? 1.25 : 1)}
        mr="xsmall"
      />
    ) : (
      renderIconPlaceholder && (
        <IconPlaceholder
          aria-hidden
          data-testid={`menu-item-${renderedIconID}-icon-placeholder`}
          mr="xsmall"
          size={24 / (compact ? 1.25 : 1)}
        />
      )
    )

  if (disabled && itemRole === 'link') {
    // eslint-disable-next-line no-console
    console.warn(
      'itemRole="link" and disabled cannot be combined - use itemRole="button" if you need to offer a disabled MenuItem'
    )
  }
  const Component = !disabled && itemRole === 'link' ? 'a' : 'button'

  const menuItemContent = (
    <Component href={href} role="menuitem" target={target}>
      {renderedIcon}
      <span>
        {children}
        {description && (
          <Paragraph color="text2" fontSize="xsmall" mt="xxsmall">
            {description}
          </Paragraph>
        )}
      </span>
      {detail && <Detail>{detail}</Detail>}
    </Component>
  )

  return (
    <MenuItemLayout
      aria-current={current && 'true'}
      compact={compact}
      disabled={disabled}
      focusVisible={isFocusVisible}
      onBlur={handleOnBlur}
      onClick={disabled ? undefined : handleOnClick}
      onKeyUp={handleOnKeyUp}
      onKeyDown={handleOnKeyDown}
      className={className}
    >
      {tooltip ? (
        <Tooltip placement={tooltipPlacement || 'left'} content={tooltip}>
          {menuItemContent}
        </Tooltip>
      ) : (
        menuItemContent
      )}
    </MenuItemLayout>
  )
}

const Detail = styled.div`
  color: ${({ theme: { colors } }) => colors.text1};
  margin-left: auto;
  margin-right: ${({ theme: { space } }) => space.medium};
  padding-left: ${({ theme: { space } }) => space.large};
`

export const MenuItem = styled(MenuItemInternal)``

interface IconPlaceholderProps extends SizeProps, SpaceProps {}

const IconPlaceholder = styled.div<IconPlaceholderProps>`
  ${size}
  ${space}
`
