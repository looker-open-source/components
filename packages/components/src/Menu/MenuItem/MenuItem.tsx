/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import merge from 'lodash/merge'
import {
  CompatibleHTMLProps,
  ColorProps,
  FontSizes,
  SpaceProps,
  TypographyProps,
} from '@looker/design-tokens'
import { IconNames } from '@looker/icons'
import React, { FC, ReactNode, useContext, useState } from 'react'
import { Icon } from '../../Icon'
import { MenuItemStyleContext } from '../MenuContext'
import { MenuItemButton } from './MenuItemButton'
import {
  MenuItemCustomization,
  MenuItemStateCustomizations,
} from './menuItemCustomization'
import { MenuItemDetail } from './MenuItemDetail'
import { MenuItemListItem } from './MenuItemListItem'
import {
  defaultMenuItemStyle,
  MenuItemStateStyle,
  MenuItemStyle,
} from './menuItemStyle'

export interface MenuSharedProps {
  customizationProps?: MenuItemCustomization
  compact?: boolean
}

// For merging compact and customizationProps from props with those from context
export function useMenuItemStyleContext(props: MenuSharedProps) {
  const { customizationProps: propCustomizations, compact: compactProp } = props
  const {
    customizationProps: contextCustomizations,
    compact: contextCompact,
  } = useContext(MenuItemStyleContext)
  const parentCustomizations = contextCustomizations || {}

  let customizationProps = parentCustomizations || propCustomizations
  if (customizationProps && parentCustomizations) {
    customizationProps = merge({}, parentCustomizations, propCustomizations)
  }

  const compact = compactProp === undefined ? contextCompact : compactProp

  return { compact, customizationProps }
}

const assignCustomizations = (
  defaultStyle: MenuItemStyle,
  changes?: MenuItemCustomization
): MenuItemStyle => {
  const { bg, color, iconColor, fontWeight, fontSize, iconSize } =
    changes || ({} as MenuItemCustomization)

  const customMarker = changes ? changes.marker : {}

  // Need to spread fontSize & iconSize across all states
  const defaults = {
    fontSize: fontSize || ('small' as FontSizes),
    iconSize: iconSize || 20,
  }

  const base: MenuItemStateCustomizations = {}

  bg && (base.bg = bg)
  color && (base.color = color)
  fontWeight && (base.fontWeight = fontWeight)
  iconColor && (base.iconColor = iconColor)

  const initial: MenuItemStateStyle = {
    ...defaultStyle.initial,
    ...defaults,
    ...base,
  }

  const current: MenuItemStateStyle = {
    ...defaultStyle.current,
    ...defaults,
    ...base,
    ...(changes ? changes.current : {}),
  }

  const hover: MenuItemStateStyle = {
    ...defaultStyle.hover,
    ...defaults,
    ...base,
    ...(changes ? changes.hover : {}),
  }

  const marker = { ...defaultStyle.marker, ...customMarker }

  return {
    current,
    hover,
    initial,
    marker,
  }
}

export interface MenuItemProps
  extends CompatibleHTMLProps<HTMLElement>,
    ColorProps,
    SpaceProps,
    TypographyProps,
    MenuSharedProps {
  detail?: ReactNode
  icon?: IconNames
  /**
   * Indicates the MenuItem is checked
   */
  current?: boolean
  /**
   * Sets the correct accessible role for the MenuItem:
   * - Use **'link'** for items that navigation to another page
   * - Use **'button'** for items that trigger in page interactions, like displaying a modal
   * @default 'button'
   *
   */
  itemRole?: 'link' | 'button'
}

export const MenuItem: FC<MenuItemProps> = (props) => {
  const {
    children,
    compact: propCompact,
    customizationProps: propCustomizations,
    current,
    detail,
    href,
    icon,
    itemRole,
    onBlur,
    onClick,
    onKeyUp,
    target,
    ...remainingProps
  } = props

  const [isFocusVisible, setFocusVisible] = useState(false)

  const handleOnKeyUp = (event: React.KeyboardEvent<HTMLLIElement>) => {
    setFocusVisible(true)
    onKeyUp && onKeyUp(event)
  }

  const handleOnBlur = (event: React.FocusEvent<HTMLLIElement>) => {
    setFocusVisible(false)
    onBlur && onBlur(event)
  }

  const handleOnClick = (event: React.MouseEvent<HTMLLIElement>) => {
    setFocusVisible(false)
    onClick && onClick(event)
  }

  const { customizationProps, compact } = useMenuItemStyleContext({
    compact: propCompact,
    customizationProps: propCustomizations,
  })
  const compactIconModifier = compact ? 1.25 : 1

  const style = assignCustomizations(defaultMenuItemStyle, customizationProps)
  const styleState = current ? style.current : style.initial
  const { iconSize, iconColor, ...listItemProps } = styleState

  const { p, py, px, pr, pl, pt, pb, ...outerProps } = remainingProps

  const clickTargetProps = {
    p,
    pb: pb || py || p || compact ? 'xxsmall' : 'small',
    pl: pl || px || p || 'medium',
    pr: pr || px || p || 'medium',
    pt: pt || py || p || compact ? 'xxsmall' : 'small',
  }

  return (
    <MenuItemListItem
      aria-current={current && 'page'}
      onClick={handleOnClick}
      current={current}
      focusVisible={isFocusVisible}
      itemStyle={style}
      onKeyUp={handleOnKeyUp}
      onBlur={handleOnBlur}
      {...listItemProps}
      {...outerProps}
    >
      <MenuItemButton
        as={itemRole === 'link' ? 'a' : 'button'}
        href={href}
        role="menuitem"
        target={target}
        {...clickTargetProps}
      >
        {icon && (
          <Icon
            name={icon}
            mr="xsmall"
            size={iconSize / compactIconModifier}
            color={iconColor}
          />
        )}
        {children}
      </MenuItemButton>
      {detail && <MenuItemDetail>{detail}</MenuItemDetail>}
    </MenuItemListItem>
  )
}
