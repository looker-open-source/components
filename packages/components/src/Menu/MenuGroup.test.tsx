/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import 'jest-styled-components'
import React from 'react'
import { assertSnapshot } from '@looker/components-test-utils'

import { palette } from '@looker/design-tokens'
import { Box } from '../Layout'
import { MenuGroup } from './MenuGroup'
import { MenuItem } from './MenuItem'

test('MenuGroup', () => {
  assertSnapshot(
    <MenuGroup>
      <MenuItem>who!</MenuItem>
    </MenuGroup>
  )
})

test('MenuGroup - label', () => {
  assertSnapshot(
    <MenuGroup label="Questions">
      <MenuItem>what?</MenuItem>
      <MenuItem>who?</MenuItem>
      <MenuItem>where?</MenuItem>
    </MenuGroup>
  )
})

test('MenuGroup - JSX label', () => {
  assertSnapshot(
    <MenuGroup label={<Box>Questions</Box>}>
      <MenuItem>what?</MenuItem>
      <MenuItem>who?</MenuItem>
      <MenuItem>where?</MenuItem>
    </MenuGroup>
  )
})

test('MenuGroup - labelProps & labelStyles', () => {
  assertSnapshot(
    <MenuGroup
      label="Questions"
      labelStyles={{ color: 'rebeccapurple' }}
      labelProps={{ fontSize: 'xxlarge' }}
    >
      <MenuItem>what?</MenuItem>
      <MenuItem>who?</MenuItem>
      <MenuItem>where?</MenuItem>
    </MenuGroup>
  )
})

test('MenuGroup - menu customization', () => {
  /* eslint-disable sort-keys */
  const menuCustomizations = {
    bg: palette.purple500,
    color: palette.purple200,
    current: {
      bg: palette.purple200,
      color: palette.purple900,
      iconColor: palette.purple500,
    },
    hover: {
      bg: palette.purple700,
      color: palette.white,
      iconColor: palette.purple100,
    },
    iconColor: palette.purple300,
    iconSize: 20,
    marker: {
      color: palette.purple900,
      size: 10,
    },
  }
  /* eslint-enable sort-keys */

  const menu = assertSnapshot(
    <MenuGroup customizationProps={menuCustomizations}>
      <MenuItem icon="LogoRings">Looker</MenuItem>
      <MenuItem icon="Validate">Validate</MenuItem>
      <MenuItem icon="ChartPie">Pizza!</MenuItem>
    </MenuGroup>
  )

  expect(menu).toMatchSnapshot()
})
