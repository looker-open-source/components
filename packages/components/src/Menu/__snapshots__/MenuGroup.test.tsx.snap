// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuGroup - JSX label 1`] = `
.c4 {
  font-family: 'Roboto','Noto Sans JP','Noto Sans CJK KR','Noto Sans Arabic UI','Noto Sans Devanagari UI','Noto Sans Hebrew','Noto Sans Thai UI','Helvetica','Arial',sans-serif;
}

.c5 {
  font-family: 'Roboto','Noto Sans JP','Noto Sans CJK KR','Noto Sans Arabic UI','Noto Sans Devanagari UI','Noto Sans Hebrew','Noto Sans Thai UI','Helvetica','Arial',sans-serif;
  list-style-type: none;
}

.c3 {
  font-family: 'Roboto','Noto Sans JP','Noto Sans CJK KR','Noto Sans Arabic UI','Noto Sans Devanagari UI','Noto Sans Hebrew','Noto Sans Thai UI','Helvetica','Arial',sans-serif;
  font-size: 0.875rem;
  font-weight: 600;
  line-height: 1.25rem;
  padding-left: 1rem;
  color: #262D33;
  color: #262D33;
}

.c1 {
  background: #FFFFFF;
  box-shadow: none;
  margin-bottom: 0.25rem;
  position: -webkit-sticky;
  position: sticky;
  top: -1px;
}

.c1 .c2 {
  color: #343C42;
}

.c0 {
  font-family: 'Roboto','Noto Sans JP','Noto Sans CJK KR','Noto Sans Arabic UI','Noto Sans Devanagari UI','Noto Sans Hebrew','Noto Sans Thai UI','Helvetica','Arial',sans-serif;
  list-style-type: none;
  padding: 0.5rem 0;
}

.c6 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #262D33;
  font-size: 0.875rem;
  font-weight: 400;
  list-style-type: none;
  outline: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition: background 150ms cubic-bezier(0.86,0,0.07,1), color 150ms cubic-bezier(0.86,0,0.07,1);
  transition: background 150ms cubic-bezier(0.86,0,0.07,1), color 150ms cubic-bezier(0.86,0,0.07,1);
}

.c6 button,
.c6 a {
  font-family: 'Roboto','Noto Sans JP','Noto Sans CJK KR','Noto Sans Arabic UI','Noto Sans Devanagari UI','Noto Sans Hebrew','Noto Sans Thai UI','Helvetica','Arial',sans-serif;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: transparent;
  border: none;
  color: inherit;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  font-size: inherit;
  font-weight: inherit;
  min-height: 40px;
  outline: none;
  padding: 0.5rem 1rem;
  position: relative;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  width: 100%;
}

.c6 button:hover,
.c6 a:hover,
.c6 button:focus,
.c6 a:focus {
  color: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c6 .c8 {
  -webkit-transition: color 150ms cubic-bezier(0.86,0,0.07,1);
  transition: color 150ms cubic-bezier(0.86,0,0.07,1);
}

.c6:hover,
.c6[aria-current='true'] {
  background: #F5F6F7;
  color: #262D33;
}

.c6[aria-current='true'] {
  font-weight: 600;
}

.c6[disabled] {
  color: #939BA5;
}

.c6[disabled] button,
.c6[disabled] a {
  cursor: not-allowed;
}

.c6[disabled]:hover {
  background: #FFFFFF;
  color: #939BA5;
}

.c7 .c8 {
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
}

<li
  className="c0"
>
  <div
    className="c1"
  >
    <div
      style={
        Object {
          "height": "0",
        }
      }
    />
    <h2
      className="c2 c3"
      color="text"
      fontFamily="body"
      fontSize="small"
      fontWeight="semiBold"
    >
      <div
        className="c4"
      >
        Questions
      </div>
    </h2>
  </div>
  <ul
    className="c5"
    type="none"
  >
    <li
      className="c6 c7"
      onBlur={[Function]}
      onClick={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
    >
      <button
        role="menuitem"
      >
        <span>
          what?
        </span>
      </button>
    </li>
    <li
      className="c6 c7"
      onBlur={[Function]}
      onClick={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
    >
      <button
        role="menuitem"
      >
        <span>
          who?
        </span>
      </button>
    </li>
    <li
      className="c6 c7"
      onBlur={[Function]}
      onClick={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
    >
      <button
        role="menuitem"
      >
        <span>
          where?
        </span>
      </button>
    </li>
  </ul>
</li>
`;

exports[`MenuGroup - label 1`] = `
.c4 {
  font-family: 'Roboto','Noto Sans JP','Noto Sans CJK KR','Noto Sans Arabic UI','Noto Sans Devanagari UI','Noto Sans Hebrew','Noto Sans Thai UI','Helvetica','Arial',sans-serif;
  list-style-type: none;
}

.c3 {
  font-family: 'Roboto','Noto Sans JP','Noto Sans CJK KR','Noto Sans Arabic UI','Noto Sans Devanagari UI','Noto Sans Hebrew','Noto Sans Thai UI','Helvetica','Arial',sans-serif;
  font-size: 0.875rem;
  font-weight: 600;
  line-height: 1.25rem;
  padding-left: 1rem;
  color: #262D33;
  color: #262D33;
}

.c1 {
  background: #FFFFFF;
  box-shadow: none;
  margin-bottom: 0.25rem;
  position: -webkit-sticky;
  position: sticky;
  top: -1px;
}

.c1 .c2 {
  color: #343C42;
}

.c0 {
  font-family: 'Roboto','Noto Sans JP','Noto Sans CJK KR','Noto Sans Arabic UI','Noto Sans Devanagari UI','Noto Sans Hebrew','Noto Sans Thai UI','Helvetica','Arial',sans-serif;
  list-style-type: none;
  padding: 0.5rem 0;
}

.c5 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #262D33;
  font-size: 0.875rem;
  font-weight: 400;
  list-style-type: none;
  outline: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition: background 150ms cubic-bezier(0.86,0,0.07,1), color 150ms cubic-bezier(0.86,0,0.07,1);
  transition: background 150ms cubic-bezier(0.86,0,0.07,1), color 150ms cubic-bezier(0.86,0,0.07,1);
}

.c5 button,
.c5 a {
  font-family: 'Roboto','Noto Sans JP','Noto Sans CJK KR','Noto Sans Arabic UI','Noto Sans Devanagari UI','Noto Sans Hebrew','Noto Sans Thai UI','Helvetica','Arial',sans-serif;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: transparent;
  border: none;
  color: inherit;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  font-size: inherit;
  font-weight: inherit;
  min-height: 40px;
  outline: none;
  padding: 0.5rem 1rem;
  position: relative;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  width: 100%;
}

.c5 button:hover,
.c5 a:hover,
.c5 button:focus,
.c5 a:focus {
  color: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c5 .c7 {
  -webkit-transition: color 150ms cubic-bezier(0.86,0,0.07,1);
  transition: color 150ms cubic-bezier(0.86,0,0.07,1);
}

.c5:hover,
.c5[aria-current='true'] {
  background: #F5F6F7;
  color: #262D33;
}

.c5[aria-current='true'] {
  font-weight: 600;
}

.c5[disabled] {
  color: #939BA5;
}

.c5[disabled] button,
.c5[disabled] a {
  cursor: not-allowed;
}

.c5[disabled]:hover {
  background: #FFFFFF;
  color: #939BA5;
}

.c6 .c7 {
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
}

<li
  className="c0"
>
  <div
    className="c1"
  >
    <div
      style={
        Object {
          "height": "0",
        }
      }
    />
    <h2
      className="c2 c3"
      color="text"
      fontFamily="body"
      fontSize="small"
      fontWeight="semiBold"
    >
      Questions
    </h2>
  </div>
  <ul
    className="c4"
    type="none"
  >
    <li
      className="c5 c6"
      onBlur={[Function]}
      onClick={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
    >
      <button
        role="menuitem"
      >
        <span>
          what?
        </span>
      </button>
    </li>
    <li
      className="c5 c6"
      onBlur={[Function]}
      onClick={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
    >
      <button
        role="menuitem"
      >
        <span>
          who?
        </span>
      </button>
    </li>
    <li
      className="c5 c6"
      onBlur={[Function]}
      onClick={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
    >
      <button
        role="menuitem"
      >
        <span>
          where?
        </span>
      </button>
    </li>
  </ul>
</li>
`;

exports[`MenuGroup 1`] = `
.c1 {
  font-family: 'Roboto','Noto Sans JP','Noto Sans CJK KR','Noto Sans Arabic UI','Noto Sans Devanagari UI','Noto Sans Hebrew','Noto Sans Thai UI','Helvetica','Arial',sans-serif;
  list-style-type: none;
}

.c0 {
  font-family: 'Roboto','Noto Sans JP','Noto Sans CJK KR','Noto Sans Arabic UI','Noto Sans Devanagari UI','Noto Sans Hebrew','Noto Sans Thai UI','Helvetica','Arial',sans-serif;
  list-style-type: none;
  padding: 0.5rem 0;
}

.c2 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #262D33;
  font-size: 0.875rem;
  font-weight: 400;
  list-style-type: none;
  outline: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition: background 150ms cubic-bezier(0.86,0,0.07,1), color 150ms cubic-bezier(0.86,0,0.07,1);
  transition: background 150ms cubic-bezier(0.86,0,0.07,1), color 150ms cubic-bezier(0.86,0,0.07,1);
}

.c2 button,
.c2 a {
  font-family: 'Roboto','Noto Sans JP','Noto Sans CJK KR','Noto Sans Arabic UI','Noto Sans Devanagari UI','Noto Sans Hebrew','Noto Sans Thai UI','Helvetica','Arial',sans-serif;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: transparent;
  border: none;
  color: inherit;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  font-size: inherit;
  font-weight: inherit;
  min-height: 40px;
  outline: none;
  padding: 0.5rem 1rem;
  position: relative;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  width: 100%;
}

.c2 button:hover,
.c2 a:hover,
.c2 button:focus,
.c2 a:focus {
  color: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c2 .c4 {
  -webkit-transition: color 150ms cubic-bezier(0.86,0,0.07,1);
  transition: color 150ms cubic-bezier(0.86,0,0.07,1);
}

.c2:hover,
.c2[aria-current='true'] {
  background: #F5F6F7;
  color: #262D33;
}

.c2[aria-current='true'] {
  font-weight: 600;
}

.c2[disabled] {
  color: #939BA5;
}

.c2[disabled] button,
.c2[disabled] a {
  cursor: not-allowed;
}

.c2[disabled]:hover {
  background: #FFFFFF;
  color: #939BA5;
}

.c3 .c4 {
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
}

<li
  className="c0"
>
  <ul
    className="c1"
    type="none"
  >
    <li
      className="c2 c3"
      onBlur={[Function]}
      onClick={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
    >
      <button
        role="menuitem"
      >
        <span>
          who!
        </span>
      </button>
    </li>
  </ul>
</li>
`;
